src/TokenRefresher.ts

```diff
@@ -0,0 +1,185 @@
+
+/**
+ * TokenRefresher class manages OAuth2 access tokens for an ioBroker adapter.
+ */
+export class TokenRefresher {
+    private readonly adapter: ioBroker.Adapter;
+    private readonly stateName: string;
+    private refreshTokenTimeout: ioBroker.Timeout | undefined;
+    private accessToken: AccessTokens | undefined;
+    private readonly url: string;
+    private readonly readyPromise: Promise<void>;
+    private readonly name: string;
+
+    /**
+     * Creates an instance of TokenRefresher.
+     *
+     * @param adapter Instance of ioBroker adapter
+     * @param serviceName Name of the service for which the tokens are managed, e.g., 'spotify', 'dropbox', etc.
+     * @param stateName Optional name of the state where tokens are stored. Defaults to 'oauth2Tokens' and that will store tokens in `ADAPTER.X.oauth2Tokens`.
+     */
+    constructor(adapter: ioBroker.Adapter, serviceName: string, stateName?: string) {
+        this.adapter = adapter;
+        this.stateName = stateName || 'oauth2Tokens';
+        this.url = `https://oauth2.iobroker.in/${serviceName}`;
+        this.name = this.stateName.replace('info.', '').replace('Tokens', '').replace('tokens', '');
+        if (this.name === 'oauth2') {
+            this.name = adapter.name;
+        }
+
+        this.readyPromise = this.adapter.getForeignStateAsync(`${adapter.namespace}.${this.stateName}`).then(state => {
+            if (state) {
+                this.accessToken = JSON.parse(state.val as string);
+                if (
+                    this.accessToken?.access_token_expires_on &&
+                    new Date(this.accessToken.access_token_expires_on).getTime() < Date.now()
+                ) {
+                    this.adapter.log.error('Access token is expired. Please make a authorization again');
+                } else {
+                    this.adapter.log.debug(`Access token for ${this.name} found`);
+                }
+            } else {
+                this.adapter.log.error(`No tokens for ${this.name} found`);
+            }
+            this.adapter
+                .subscribeStatesAsync(this.stateName)
+                .catch(error => this.adapter.log.error(`Cannot read tokens: ${error}`));
+
+            return this.refreshTokens().catch(error => this.adapter.log.error(`Cannot refresh tokens: ${error}`));
+        });
+    }
+
+    /**
+     * Destroys the TokenRefresher instance, clearing any timeouts and stopping state subscriptions.
+     */
+    destroy(): void {
+        if (this.refreshTokenTimeout) {
+            this.adapter.clearTimeout(this.refreshTokenTimeout);
+            this.refreshTokenTimeout = undefined;
+        }
+    }
+
+    /**
+     * This method is called when the state changes for the token.
+     *
+     * @param id ID of the state that changed
+     * @param state Value
+     */
+    onStateChange(id: string, state: ioBroker.State | null | undefined): void {
+        if (state?.ack && id.endsWith(`.${this.stateName}`)) {
```

Comment:
I would do it a bit more exact ... because you know exactly the path ... otherwise you might react on wrong states